src/iterator.cr

```diff
@@ -131,13 +131,19 @@ module Iterator(T)
   end
 
   def self.of(&block : -> T)
-    SingletonProc(T).new(block)
+    SingletonProc(typeof(without_stop(&block))).new(block)
+  end
+
+  private def self.without_stop(&block : -> T)
+    e = block.call
+    raise "" if e.is_a?(Iterator::Stop)
```

Comment:
We just need a function that returns `NoReturn`. The same `raise ""` pattern is used in https://github.com/crystal-lang/crystal/blob/master/src/iterator.cr#L472 .
The code is never executed actually.